From 05c5fac0f8276334a14b24454282a74f023e9ea3 Mon Sep 17 00:00:00 2001
From: Anson Huang <b20788@freescale.com>
Date: Thu, 19 Jun 2014 10:36:47 +0800
Subject: [PATCH 0304/1543] ENGR00318913-7 ARM: imx: add cpufreq support for
 i.mx6sx

commit 438e9fa90f80a5c4fbfd7282796ca514823afac6 from
git://git.freescale.com/imx/linux-2.6-imx.git

Enable cpufreq support for i.MX6SX, currently three setpoints
are supported, the freq/volt table are as below:

        VDDARM_CAP      VDDSOC_CAP
996M:   1.250V          1.175V
792M:   1.175V          1.175V
396M:   1.075V          1.175V

All upper voltages are 25mV higher then the minimum value defined
in datasheet, this 25mV is to cover board level IR drop.

Signed-off-by: Anson Huang <b20788@freescale.com>
---
 arch/arm/mach-imx/mach-imx6sx.c |   33 ++++++++++++++++++++++++++++++++-
 1 files changed, 32 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-imx/mach-imx6sx.c b/arch/arm/mach-imx/mach-imx6sx.c
index 02fccf6..95f114d 100644
--- a/arch/arm/mach-imx/mach-imx6sx.c
+++ b/arch/arm/mach-imx/mach-imx6sx.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2014 Freescale Semiconductor, Inc.
+ * Copyright (C) 2014 Freescale Semiconductor, Inc.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 as
@@ -8,6 +8,7 @@
 
 #include <linux/irqchip.h>
 #include <linux/of_platform.h>
+#include <linux/pm_opp.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/map.h>
 
@@ -42,10 +43,40 @@ static const char *imx6sx_dt_compat[] __initconst = {
 	NULL,
 };
 
+static void __init imx6sx_opp_init(struct device *cpu_dev)
+{
+	struct device_node *np;
+
+	np = of_find_node_by_path("/cpus/cpu@0");
+	if (!np) {
+		pr_warn("failed to find cpu0 node\n");
+		return;
+	}
+
+	cpu_dev->of_node = np;
+	if (of_init_opp_table(cpu_dev))
+		pr_warn("failed to init OPP table\n");
+
+	of_node_put(np);
+}
+
+static struct platform_device imx6sx_cpufreq_pdev = {
+	.name = "imx6q-cpufreq",
+};
+
+static void __init imx6sx_init_late(void)
+{
+	if (IS_ENABLED(CONFIG_ARM_IMX6Q_CPUFREQ)) {
+		imx6sx_opp_init(&imx6sx_cpufreq_pdev.dev);
+		platform_device_register(&imx6sx_cpufreq_pdev);
+	}
+}
+
 DT_MACHINE_START(IMX6SX, "Freescale i.MX6 SoloX (Device Tree)")
 	.map_io		= debug_ll_io_init,
 	.init_irq	= imx6sx_init_irq,
 	.init_machine	= imx6sx_init_machine,
+	.init_late	= imx6sx_init_late,
 	.dt_compat	= imx6sx_dt_compat,
 	.restart	= mxc_restart,
 MACHINE_END
-- 
1.7.5.4

